### PR TITLE
Integration specs and N + 1 problems

### PR DESCRIPTION
**Pull request by mikemtzp**
**N+1**
- All problems are solved when fetching all posts and their comments for a user using the `bullet` gem.

**Integration specs**
- Use `capybara` to write integration tests for each view.

- User index page:
  - I can see the username of all other users.
  - I can see the profile picture for each user.
  - I can see the number of posts each user has written.
  - When I click on a user, I am redirected to that user's show page.
 
- User show page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see the user's bio.
  - I can see the user's first 3 posts.
  - I can see a button that lets me view all of a user's posts.
  - When I click a user's post, it redirects me to that post's show page.
  - When I click to see all posts, it redirects me to the user's post's index page.
 
- User post index page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see a post's title.
  - I can see some of the post's body.
  - I can see the first comments on a post.
  - I can see how many comments a post has.
  - I can see how many likes a post has.
  - When I click on a post, it redirects me to that post's show page.

- Post show page:
  - I can see the post's title.
  - I can see who wrote the post.
  - I can see how many comments it has.
  - I can see how many likes it has.
  - I can see the post body.
  - I can see the username of each commentor.
  - I can see the comment each commentor left.

### CODE REVIEW

File: spec/features/post_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  end
+
+  it 'should show how many comments a post has' do
+    expect(page).to have_content(post.comments_counter)
+  end
+
+  it 'should show how many likes a post has' do
+    expect(page).to have_content(post.likes_counter)
+  end
+
```
Comment (danmainah) [inline spec/features/post_index_spec.rb:47]: - [x] @mikemtzp and @Daf98 great work on your test you forgot to add one test for checking for pagination if there are more posts than fit on the view.

File: spec/features/post_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
+require 'rails_helper'
+
+RSpec.describe 'post_index', type: :feature do
+  user = User.find_by(name: 'Tom')
+  post = Post.where(author_id: user.id).last
```
Comment (danmainah) [inline spec/features/post_index_spec.rb:5]: - [x] Kindy check your test, I run into the error shown below when running your tests:

image

File: spec/features/post_show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
+require 'rails_helper'
+
+RSpec.describe 'post_show', type: :feature do
```
Comment (danmainah) [inline spec/features/post_show_spec.rb:3]: - [x] Kindly check on your test. I run into the error below while running your tests.
image

File: spec/features/user_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
+require 'rails_helper'
+
+RSpec.describe 'user_show', type: :feature do
```
Comment (danmainah) [inline spec/features/user_show_spec.rb:3]: - [x] Kindly check on your test. I run into the error below while running your tests.

image

- [x] Kindy ensure all the tests are passing.

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @mikemtzp and @Daf98 , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

✔️ No linter errors
✔️ Github flow was correctly used


# Required Changes ♻️

- [x] Kindy ensure you optimize your controllers to solve the N+1 problem when fetching all posts and their comments.

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Comment by mikemtzp** [comment]
Hi, @danmainah thanks for the feedback! 😁 

- Regarding the N + 1 optimization my app asks me to avoid using `includes` as it is being detected as `eager loading` in the post controller.



- However, I added the following on the `posts/index.html.erb` file to improve performance:

```
post.recent_comments.includes(:author).each do |comment| %>
          <li class="comment"><%= comment.author.name %>: <%= comment.text %></li>
```

- I also created the pagination test. All the tests are passing on my side.

**Review by emmanuellekamwa — APPROVED** [review]
**STATUS: APPROVED 🍾✔🎉**
Hello @mikemtzp 👋
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

- You documented your work professionally.✔
- Your linters are passing all the checks
- You asked for the changes in a polite and friendly way.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**